### PR TITLE
[Monitoring] Generify voltage and angle monitoring 

### LINKED ISSUE #899: Merge voltage & angle monitoring algorithms
### Describe the current behavior

Voltage & Angle monitoring algorithms are pretty similar, but are handled in different classes

### Describe the expected behavior

Merge them into one generic configurable algorithm

### Describe the motivation

Cleaner code and less duplication

### Extra Information

_No response_

### PR DESCRIPTION
**Pull request by benrejebmoh**
**Please check if the PR fulfills these requirements**

- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

Fixes #899 

**What kind of change does this PR introduce?**

**What is the current behavior?**

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**

- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**

**Other information**:

### CODE REVIEW

File: generic-monitoring/pom.xml
Comment (pet-mit) [inline generic-monitoring/pom.xml:1]: I would remove "generic" from the name

File: generic-monitoring/src/main/java/com/powsybl/openrao/monitoring/MonitoringInput.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            this.scalableZonalData = scalableZonalData;
+            return this;
+        }
+
+        public MonitoringInput build() {
+            MonitoringInput monitoringInput = new MonitoringInput();
+            monitoringInput.crac = Objects.requireNonNull(crac, format(REQUIRED_ARGUMENT_MESSAGE, "CRAC"));
+            monitoringInput.network = Objects.requireNonNull(network, format(REQUIRED_ARGUMENT_MESSAGE, "Network"));
+            monitoringInput.raoResult = Objects.requireNonNull(raoResult, format(REQUIRED_ARGUMENT_MESSAGE, "RaoResult"));
+            monitoringInput.physicalParameter = physicalParameter;
```
Comment (pet-mit) [inline generic-monitoring/src/main/java/com/powsybl/openrao/monitoring/MonitoringInput.java:57]: what happens if pyhsicalParameter is null?

File: generic-monitoring/pom.xml
Comment (benrejebmoh) [reply generic-monitoring/pom.xml:1]: i put 'generic' temporarily in the name of the module to keep in the same branch the old monitoring module until the validation of this PR

File: generic-monitoring/src/main/java/com/powsybl/openrao/monitoring/Monitoring.java [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+                // For Cnecs with overshoot, get associated remedial actions
+                Set<NetworkAction> availableNetworkActions = getNetworkActionsAssociatedToCnec(state, monitoringInput.getCrac(), cnecResult.getCnec(), monitoringInput.getPhysicalParameter());
+                // and apply them
+                AppliedNetworkActionsResult appliedNetworkActionsResult = applyNetworkActions(network, availableNetworkActions, cnecResult.getCnec().getId(), monitoringInput);
+                appliedNetworkActionsResultList.add(appliedNetworkActionsResult);
+            }
+        });
+
+        // if AngleMonitoring : Redispatch to compensate the loss of generation/ load
+        if (monitoringInput.getPhysicalParameter().equals(PhysicalParameter.ANGLE)) {
```
Comment (pet-mit) [inline generic-monitoring/src/main/java/com/powsybl/openrao/monitoring/Monitoring.java:158]: not necessary. can be generic if we only check if any of the applied RAs are InjectionSetpoints.
more generally, I think we can make it cleaner / more generic by doing a "balnce" check before & after the monitoring

File: generic-monitoring/src/main/java/com/powsybl/openrao/monitoring/AppliedNetworkActionsResult.java
Comment (pet-mit) [inline generic-monitoring/src/main/java/com/powsybl/openrao/monitoring/AppliedNetworkActionsResult.java:1]: I'd prefer if we can remove this class

**Review by pet-mit — CHANGES_REQUESTED** [review]

**Review by pet-mit — DISMISSED** [review]

**Review by Godelaine — CHANGES_REQUESTED** [review]

File: data/crac/crac-impl/src/main/java/com/powsybl/openrao/data/cracimpl/FlowCnecImpl.java [hunk 3]
```diff
@@ -156,6 +154,89 @@ public PhysicalParameter getPhysicalParameter() {
...
+        } else {
+            power1 = getPower(branch, TwoSides.ONE, unit);
+            result.put(TwoSides.ONE, power1);
+            return new FlowCnecValue(power1, Double.NaN);
+        }
+    }
+
+    private double getPower(Branch branch, TwoSides side, Unit unit) {
+        double power = unit == Unit.MEGAWATT ? branch.getTerminal(side).getP() : branch.getTerminal(side).getI();
+        return Double.isNaN(power) ? branch.getTerminal(side).getP() * getFlowUnitMultiplier(side) : power;
```
[the remark this answers is not in the document]
Comment (benrejebmoh) [reply data/crac/crac-impl/src/main/java/com/powsybl/openrao/data/cracimpl/FlowCnecImpl.java:183]: i am not sure to understand "explicit that you're doing a conversion Ampere/MW"

File: data/crac/crac-impl/src/test/java/com/powsybl/openrao/data/cracimpl/FlowCnecImplTest.java [hunk 4]
```diff
@@ -63,6 +64,79 @@ void testGetLocation1() {
         assertTrue(countries.contains(Optional.of(Country.NL)));
     }
 
+    @Test
+    void testComputeValue() {
+        Network network = Mockito.mock(Network.class, Mockito.RETURNS_DEEP_STUBS);
```
[the remark this answers is not in the document]
Comment (benrejebmoh) [reply data/crac/crac-impl/src/test/java/com/powsybl/openrao/data/cracimpl/FlowCnecImplTest.java:69]: I mocked in purpose the branches to check that computeValue method in FlowCnecImpl calculates correctly side1Value and side2Value. In the TestCase12Nodes.uct network  P on terminal is Nan.

[the remark this answers is not in the document]
Comment (Godelaine) [reply data/crac/crac-impl/src/test/java/com/powsybl/openrao/data/cracimpl/FlowCnecImplTest.java:69]: ok, never mind then, could you still remove RETURNS_DEEP_STUBS even if that means a few more lines to mock terminals ?

**Review by Godelaine — APPROVED** [review]

**Pull request merged**